### PR TITLE
Modified MarkupTokenizer, so that escaped markup inside markup is valid

### LINKED ISSUE #903: EscapeMarkup doesn't work when using [link=...]
**Information**
 - OS: [Windows]
 - Version: [0.44.0]
 - Terminal: [Windows Terminal]

**Describe the bug**
Escape doesn't work in some cases. It seems in combination with [link] it throws an exception `Encountered unescaped ']' token at position 56`

**To Reproduce**
```csharp
            var path = "C:\\Dev\\ExampleProjects\\TestApp\\TestApp[-].UnitTests\\StrykerOutput\\2022-07-22.15-27-40\\reports\\mutation-report.html";
            // the next line prints our path
            AnsiConsole.Console.MarkupLineInterpolated($"[Green]{path}[/]");
            // the next line throws an exception
            AnsiConsole.Console.MarkupLineInterpolated($"[Green][link={path}]{path}[/][/]");
```
**Expected behavior**
No exception.

**Additional context**
This is an actual use case in Stryker.

### PR DESCRIPTION
**Pull request by nils-a**
fixes #903

### CODE REVIEW

**Review by phil-scott-78 — APPROVED** [review]
Been a minute since I've looked in here, but things look good to me. 

Maybe it has been my time away from digging into the guts, but if you have to jump back in to fix the formatting issue it might be nice to comment out the difference scenarios we are checking for in the tokenizer. The nesting is getting kind of deep and it was tough for my admittedly sleep deprived brain to follow :-)

**Comment by nils-a** [comment]
Yes - it was kind of late yesterday, but I wanted to get the code out, so you could pick on it 😁 
I'm not really fond of the nesting either - I had a hard time grasping what was going on. Then again, I was never good at writing parsers. 🤷

File: src/Spectre.Console/Internal/Text/Markup/MarkupTokenizer.cs [hunk 1]
```diff
@@ -113,7 +138,7 @@ public bool MoveNext()
                     if (encounteredClosing)
                     {
                         throw new InvalidOperationException(
-                            $"Encountered unescaped ']' token at position {_reader.Position}");
+                            $"Encountered unescaped ']' token at position {_reader.Position}.");
```
Comment (patriksvensson) [inline src/Spectre.Console/Internal/Text/Markup/MarkupTokenizer.cs:141]: I understand the tempation to add the dot here, but we should remove it so we're consistent with the rest of the library (or change all of them) 😅

File: src/Spectre.Console/Internal/Text/Markup/MarkupTokenizer.cs [hunk 2]
```diff
@@ -66,21 +66,46 @@ public bool MoveNext()
...
+                }
+
+                if (next == '[')
+                {
+                    _reader.Read();
+                }
+                else
+                {
+                    throw new InvalidOperationException(
+                        $"Encountered malformed markup tag at position {_reader.Position - 1}.");
```
Comment (patriksvensson) [inline src/Spectre.Console/Internal/Text/Markup/MarkupTokenizer.cs:100]: See comment about punctuation.

File: src/Spectre.Console/Internal/Text/Markup/MarkupTokenizer.cs [hunk 3]
```diff
@@ -122,7 +147,7 @@ public bool MoveNext()
 
             if (encounteredClosing)
             {
-                throw new InvalidOperationException($"Encountered unescaped ']' token at position {_reader.Position}");
+                throw new InvalidOperationException($"Encountered unescaped ']' token at position {_reader.Position}.");
```
Comment (patriksvensson) [inline src/Spectre.Console/Internal/Text/Markup/MarkupTokenizer.cs:150]: See comment about punctuation.

**Review by patriksvensson — CHANGES_REQUESTED** [review]

File: src/Spectre.Console/Internal/Text/Markup/MarkupTokenizer.cs (hunk 1 above)
Comment (nils-a) [reply src/Spectre.Console/Internal/Text/Markup/MarkupTokenizer.cs]: I've reverted that change. I started to make them consistent but stopped - that is a far too big change to slide it in here 🤣

**Comment by nils-a** [comment]
I've restructured the `MarkupTokenizer` a bit so that it's hopefully a bit more readable/understandable.

I let the text of the exceptions basically as they were before this PR - I'll start a PR to make the exceptions all have the "same" punctuation as soon as this one's fine.

**Comment by rouke-broersma** [comment]
Thank you for fixing this! Looking forward to a release :)

**Review by patriksvensson — APPROVED** [review]

**Pull request merged**